[RDY] Generate PARAM_COUNT macro

glacambre opened the pull request:
This PR fixes a bug where running `:echo nvim_get_all_options_info()` would crash.

jamessan commented:
Why wasn't `ARRAY_SIZE()` working?

glacambre commented:
Because `ARRAY_SIZE` counted the `={.fullname=NULL}` at the end of the array as an option - which resulted in neovim attempting to serialize it.

glacambre commented:
@jamessan Done, I updated the commit :)

jamessan commented:
Why the change to Lua script rather than

```
// The last element is a sentinel value, so don't include it in the total count
#define PARAM_COUNT ARRAY_SIZE(options) - 1
```

or having the serialize handle the NULL?

glacambre commented:
@jamessan Adding a `-1` was my first reaction too, but then I wanted to add a comment to the piece of code that generated the options array to warn that the PARAM_COUNT macro would need to be updated if the `.fullname=NULL` poison was removed and then realized that it'd make more sense to have the size of the array be generated by the piece of code that generates the array.

>having the serialize handle the NULL?

This would mean removing an assert, which I don't think is a good idea.

Moving the generation of PARAM_COUNT to the script that generates the array feels like the best solution to me but you seem to disagree, is there a reason for that? :)

jamessan commented:
I'm not strictly opposed to changing the lua script.

I think the more relevant issue is that we have mixed usages for determining whether we've reached the end.  We have some code looping while `opt[i].fullname != NULL` and other code using `PARAM_COUNT` as a bound.  Rather than unifying on one approach, we're just fixing the latter to avoid incorrectly accessing a NULL pointer.

glacambre commented:
>I think the more relevant issue is that we have mixed usages for determining whether we've reached the end.

Ah, I hadn't noticed that! I think the best approach would be to update the code that relies on `.fullname != NULL` to use PARAM_COUNT instead - do you agree? If yes I'll perform update, it looks there are just 3 loops that need to be changed.

jamessan commented:
Any preference, @bfredl?

glacambre committed:
Fix crash on `:echo get_all_options_info()`

Iterating over PARAM_COUNT is wrong as PARAM_COUNT also counts the last
element of the options array, which has a NULL fullname in order to
signal the end of the array.

bfredl commented:
We were planning to change to `fullname != NULL` in the vim.opt PR. The other way also works but then i would also rename the macro to indicate it has been changed when porting patches.

glacambre commented:
Changing to `fullname != NULL` makes the diff much smaller, which will probably make porting future patches from Vim easier, so it's a much better idea than my proposed solution.

@jamessan In the end that's probably what you suggested when you said "having the serialize handle the NULL?" but I misunderstood that as a suggestion for a lower-level change, sorry about that 🤦.
